imap_processing/hi/l1a/histogram.py

```diff
@@ -0,0 +1,122 @@
+    # TODO: Confirm whether to define bins by centers or edges. For now centers
+    #    are assumed.
+    coords["angle"] = xr.DataArray(
+        np.arange(2, 360, 4),
+        name="angle",
+        dims=["angle"],
+        attrs=hi_cdf_attrs.hi_hist_l1a_angle_attrs.output(),
+    )
+
+    data_vars = dict()
+    data_vars["esa_step"] = xr.DataArray(
+        np.empty(len(packets), np.dtype("uint8")),
+        dims=["epoch"],
+        attrs=dataclasses.replace(
+            hi_cdf_attrs.hi_hist_l1a_esa_step_attrs,
+            catdesc=packets[0].data["ESA_STEP"].long_description,
+            fieldname="ESA_STEP",
+            label_axis="ESA_STEP",
+        ).output(),
+    )
+
+    data_vars["acq_start"] = xr.DataArray(
+        np.empty(len(packets), np.dtype("uint16")),
+        dims=["epoch"],
+        attrs=dataclasses.replace(
+            hi_cdf_attrs.hi_hist_l1a_acq_start_attrs,
+            catdesc=packets[0].data["ACQ_START"].long_description,
+            fieldname="ACQ_START",
+            label_axis="ACQ_START",
+        ).output(),
+    )
+
+    # This dictionary defines the names and order of the 24, 90 element counter arrays
+    # TODO: This implementation will result in a CDF with variables consisting of the
+    #    counter_def_dict key + tuple string. Need to verify if this is OK with
+    #    instrument team
+    counter_def_dict = {
+        "QUAL": ("AB", "C1C2", "AC1", "BC1", "ABC1", "AC1C2", "BC1C2", "ABC1C2"),
+        "LONG": (
+            "A",
+            "B",
+            "C",
+            "AB",
+            "C1C2",
+            "AC1",
+            "BC1",
+            "ABC1",
+            "AC1C2",
+            "BC1C2",
+            "ABC1C2",
+        ),
+        "MISC": ("TOTAL_A", "TOTAL_B", "TOTAL_C", "FEE_DE_SENT", "FEE_DE_RECD"),
+    }
+    for key, val in counter_def_dict.items():
+        for counter in val:
+            data_vars[f"{key}_{counter}"] = xr.DataArray(
+                data=np.empty((len(packets), len(coords["angle"])), np.dtype("uint16")),
+                dims=["epoch", "angle"],
+                attrs=dataclasses.replace(
+                    hi_cdf_attrs.hi_hist_l1a_counter_attrs,
+                    catdesc=f"{key}_{counter} Counter",
+                    fieldname=f"{key}_{counter}",
+                    label_axis=f"{key}_{counter}",
+                ).output(),
+            )
+
+    dataset = xr.Dataset(
+        data_vars=data_vars,
+        coords=coords,
+        attrs=hi_cdf_attrs.hi_hist_l1a_global_attrs.output(),
+    )
+
+    # unpack the packets data into the Dataset
+    for i_epoch, packet in enumerate(packets):
+        dataset.epoch.data[i_epoch] = calc_start_time(
+            packet.data["CCSDS_MET"].raw_value
+        )
+        dataset.esa_step.data[i_epoch] = packet.data["ESA_STEP"].raw_value
+        dataset.acq_start.data[i_epoch] = packet.data["ACQ_START"].raw_value
+
+        counters = ConstBitStream(bin=packet.data["COUNTERS"].raw_value)
```

Comment:
I've been trying to limit our usage of bitstring when possible for performance reasons. Are we able to convert these values ourselves with native Python functionality? Here is a quick previous example: https://github.com/IMAP-Science-Operations-Center/imap_processing/pull/309

Reply:
Would this work to get a list of ints from that data? This doesn't seem too bad to me.
```python
data_string = packet.data["COUNTERS"].raw_value
list_of_ints = [int(data_string[i*12:(i+1)*12], 2) for i in range(90*24)]
```

Profiling it is 10x faster:
```
❯ python -m timeit 'import bitstring; x = "10" * 6 * 90 * 24; counters = bitstring.ConstBitStream(bin=x); data = counters.readlist("2160*uint12")'
50 loops, best of 5: 7.66 msec per loop
❯ python -m timeit 'import bitstring; x = "10" * 6 * 90 * 24; counters = bitstring.ConstBitStream(bin=x); data = [int(x[i*12:(i+1)*12], 2) for i in range(90*24)]'
500 loops, best of 5: 742 usec per loop
```

Reply:
Wow, 10x faster!
I have essentially the same implementation staged to commit. I am working through a few other changes in this PR as well.

Reply:
I am amazed by solutions you guys come up with. Nice!